app/src/main/java/org/astraea/app/common/Cache.java

```diff
@@ -0,0 +1,156 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.astraea.app.common;
+
+import static java.util.Objects.requireNonNull;
+
+import java.time.Duration;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+public class Cache<K, V> {
+
+  public interface CacheLoader<K, V> {
+    V load(K key);
+  }
+
+  public interface RemovalListener<K, V> {
+    void onRemove(Map.Entry<K, V> entry);
+  }
+
+  public static CacheBuilder<Object, Object> builder() {
+    return new CacheBuilder<>();
+  }
+
+  public static final class CacheBuilder<A, B> {
+    private CacheLoader<? super A, B> cacheLoader;
+    private long expireAfterAccessNanos;
+    private RemovalListener<? super A, ? super B> removalListener;
+    private int maxCapacity = -1;
+
+    public static CacheBuilder<Object, Object> builder() {
+      return new CacheBuilder<>();
+    }
+
+    private CacheBuilder() {}
+
+    public CacheBuilder<A, B> expireAfterAccess(Duration duration) {
+      this.expireAfterAccessNanos = duration.toNanos();
+      return this;
+    }
+
+    public CacheBuilder<A, B> maxCapacity(int maxCapacity) {
+      this.maxCapacity = maxCapacity;
+      return this;
+    }
+
+    public <A1 extends A, B1 extends B> CacheBuilder<A1, B1> removalListener(RemovalListener<?
+        super A1, ? super B1> removalListener) {
+      @SuppressWarnings("unchecked")
+      CacheBuilder<A1, B1> me = (CacheBuilder<A1, B1>) this;
+      me.removalListener = requireNonNull(removalListener);
+      return me;
+    }
+
+    public <A1 extends A, B1 extends B> CacheBuilder<A1, B1> cacheLoader(CacheLoader<? super A1,
+        B1> cacheLoader) {
+      @SuppressWarnings("unchecked")
+      CacheBuilder<A1, B1> me = (CacheBuilder<A1, B1>) this;
```

Comment:
This casting is dangerous since the type may be incompatible to type injected by `removalListener`. For example, the following test can't pass.

```java
  @Test
  void test() {
    var cache = Cache.builder()
            .cacheLoader((Base k) -> new Base())
            .removalListener((Map.Entry<Sub, Sub> e) -> {})
            .build();
    var a = cache.get(new Sub());
  }

  class Base {
  }

  class Sub extends Base {
  }
```

Reply:
Thanks for the catch, removalListener should always happen before cacheLoader, to force user follow this order, I'll move cacheLoader inside build().